source/simulator/assembly.cc

```diff
@@ -1570,13 +1588,15 @@ namespace aspect
                       // We only need to look up values of shape functions if they
                       // belong to 'our' component. They are zero otherwise anyway.
                       // Note that we later only look at the values that we do set here.
-                      for (unsigned int k=0; k<advection_dofs_per_cell; ++k)
-                        {
-                          scratch.face_grad_phi_field[k]          = (*scratch.face_finite_element_values)[solution_field].gradient (scratch.face_finite_element_values->get_fe().component_to_system_index(solution_component, k), q);
-                          scratch.face_phi_field[k]               = (*scratch.face_finite_element_values)[solution_field].value (scratch.face_finite_element_values->get_fe().component_to_system_index(solution_component, k), q);
-                          scratch.neighbor_face_grad_phi_field[k] = (*scratch.neighbor_face_finite_element_values)[solution_field].gradient (scratch.neighbor_face_finite_element_values->get_fe().component_to_system_index(solution_component, k), q);
-                          scratch.neighbor_face_phi_field[k]      = (*scratch.neighbor_face_finite_element_values)[solution_field].value (scratch.neighbor_face_finite_element_values->get_fe().component_to_system_index(solution_component, k), q);
-                        }
+                      for (unsigned int k=0, i=0; i<advection_dofs_per_cell; ++k)
+                        if (fe.system_to_component_index(k).first == solution_component)
```

Comment:
Since all this are local indices (to the element) I would hope they all number in the same way. Do we currently support any case where we have different elements in different parts of the domain?